dhall-lsp-server/src/LSP/Handlers/Diagnostics.hs

```diff
@@ -23,34 +22,56 @@ import Control.Monad.Reader (ReaderT)
 import Control.Monad.Reader.Class (ask)
 import Control.Monad.Trans (lift, liftIO)
 
+import Data.Text (Text)
+
 import Backend.Dhall.Diagnostics
 
 
 
 -- TODO: Make max number of errors parameter configurable (not rly relevant since we got 1, but still) 
 -- ---------------------------------------------------------------------
 -- Y no method to flush particular source errors?
-sendEmptyDiagnostics ::  J.Uri -> Maybe Int -> ReaderT (LSP.Core.LspFuncs ()) IO ()
+sendEmptyDiagnostics ::  LSP.Uri -> Maybe Int -> ReaderT (LSP.Core.LspFuncs ()) IO ()
 sendEmptyDiagnostics fileUri version =
   publishDiagnostics 10 fileUri version defaultDiagnosticBySource
 
 defaultDiagnosticBySource :: DiagnosticsBySource
-defaultDiagnosticBySource = Map.singleton (Just defaultDiagnosticSource) (Data.SortedList.toSortedList [])
+defaultDiagnosticBySource = Map.singleton (Just "Dhall") (Data.SortedList.toSortedList [])
+
+diagnosisToLSP :: Diagnosis -> LSP.Diagnostic
+diagnosisToLSP Diagnosis{..} = LSP.Diagnostic {..}
+  where
+    _range = case range of
+      Just (Range (line1, col1) (line2, col2)) ->
+        LSP.Range (LSP.Position line1 col1) (LSP.Position line2 col2)
+      Nothing -> LSP.Range (LSP.Position 0 0) (LSP.Position 0 0)
+    _severity = Just LSP.DsError
+    -- TODO use @doctor@ instead. Seems incompatible with LSP.Core's API...
+    _source = Just "Dhall"
+    _code = Nothing
+    _message = diagnosis
+    _relatedInformation = Nothing
 
+
+compilerDiagnostics :: FilePath -> Text -> IO [LSP.Diagnostic]
+compilerDiagnostics path txt = do
+  errors <- runDhall path txt
+  let diagnoses = concatMap (diagnose txt) errors
+  return (map diagnosisToLSP diagnoses)
+  
 -- | Analyze the file and send any diagnostics to the client in a
 -- "textDocument/publishDiagnostics" notification
-
-sendDiagnostics :: J.Uri -> Maybe Int ->  ReaderT (LSP.Core.LspFuncs ()) IO ()
+sendDiagnostics :: LSP.Uri -> Maybe Int ->  ReaderT (LSP.Core.LspFuncs ()) IO ()
 sendDiagnostics fileUri version = do
   let
-   filePath = maybe (error "can't convert uri to file path") id $ J.uriToFilePath fileUri -- !FIXME: handle non-file uris
+   filePath = maybe (error "can't convert uri to file path") id $ LSP.uriToFilePath fileUri -- !FIXME: handle non-file uris
```

Comment:
I also try to avoid the use of `error` and prefer using `fail` (specialized to `ReaderT … IO` in this case) instead:

```haskell
filePath <- case LSP.uriToFilePath of
    Nothing       -> fail "can't convert uri to fail path"
    Just filePath -> return filePath
```

The main reason why is that `error` is sensitive to evaluation order, whereas `fail` is not (at least, for `IO`).